btree/btree.c

```diff
@@ -0,0 +1,13037 @@
+ * If a node descriptor pointing to this node is already loaded in memory then
+ * this function will increment the reference count in the node descriptor
+ * before returning it to the caller.
+ * If the parameter tree is NULL then the function assumes the node at segaddr
+ * to be the root node and will also load the tree descriptor in memory for
+ * this root node.
+ *
+ * @param op load operation to perform.
+ * @param tree pointer to tree whose node is to be loaded or NULL if tree has
+ *             not been loaded.
+ * @param addr node address in the segment.
+ * @param nxt state to return on successful completion
+ *
+ * @return next state
+ */
+static int64_t bnode_get(struct node_op *op, struct td *tree,
+			 struct segaddr *addr, int nxt)
+{
+	const struct node_type *nt;
+	struct nd              *node;
+	bool                    in_lrulist;
+	uint32_t                ntype;
+
+	/**
+	 * TODO: Include function bnode_access() during async mode of btree
+	 * operations to ensure that the node data is loaded form the segment.
+	 * Also consider adding a state here to return as we might need some
+	 * time to load the node if it is not loaded.
+	 */
+
+	/**
+	 * TODO: Adding list_lock to protect from multiple threads
+	 * accessing the same node descriptor concurrently.
+	 * Replace it with a different global lock once hash
+	 * functionality is implemented.
+	 */
+
+	m0_rwlock_write_lock(&list_lock);
+
+	/**
+	 * If the node was deleted before node_get can acquire list_lock, then
+	 * restart the tick funcions.
+	 */
+	if (!segaddr_header_isvalid(addr)) {
+		op->no_op.o_sm.sm_rc = M0_ERR(-EINVAL);
+		m0_rwlock_write_unlock(&list_lock);
+		return nxt;
+	}
+	ntype = segaddr_ntype_get(addr);
+	nt = btree_node_format[ntype];
+
+	op->no_node = nt->nt_opaque_get(addr);
+
+	if (op->no_node != NULL &&
+	    op->no_node->n_addr.as_core == addr->as_core) {
+
+		bnode_lock(op->no_node);
+		if (!op->no_node->n_be_node_valid) {
+			op->no_op.o_sm.sm_rc = M0_ERR(-EACCES);
+			bnode_unlock(op->no_node);
+			m0_rwlock_write_unlock(&list_lock);
+			return nxt;
+		}
+
+		in_lrulist = op->no_node->n_ref == 0;
+		op->no_node->n_ref++;
+		if (in_lrulist) {
+			/**
+			 * The node descriptor is in LRU list. Remove from lru
+			 * list and add to active list.
+			 */
+			ndlist_tlist_del(op->no_node);
+			ndlist_tlist_add(&btree_active_nds, op->no_node);
+			lru_space_used -= (m0_be_chunk_header_size() +
+					   op->no_node->n_size);
+			/**
+			 * Update nd::n_tree  to point to tree descriptor as we
+			 * as we had set it to NULL in bnode_put(). For more
+			 * details Refer comment in bnode_put().
+			 */
+			op->no_node->n_tree = tree;
+		}
+		bnode_unlock(op->no_node);
+	} else {
+		/**
+		 * If node descriptor is already allocated for the node, no need
+		 * to allocate node descriptor again.
+		 */
+		op->no_node = nt->nt_opaque_get(addr);
+		if (op->no_node != NULL &&
+		    op->no_node->n_addr.as_core == addr->as_core) {
```

Comment:
It seems the code on line 2058-61 is doing similar operation, is it needed ?
op->no_node = nt->nt_opaque_get(addr); ....